File: rt-host/teleop/gripper_cmd.cpp
// gripper_cmd.cpp — One-shot gripper command (open/close/grasp)
// Usage:
//   gripper_cmd <ip> open                       # open to max
//   gripper_cmd <ip> close                      # close to 0
//   gripper_cmd <ip> width <w_m> [--speed 0.1]  # move to the specified width (meters)
//   gripper_cmd <ip> grasp <w_m> [--speed 0.1] [--force 30]
//   gripper_cmd <ip> homing                     # calibrate (~10s)
//   gripper_cmd <ip> read                       # print state
#include <franka/exception.h>
#include <franka/gripper.h>
#include <iostream>
#include <string>

int main(int argc, char** argv) {
  if (argc < 3) {
    std::cerr << "Usage: " << argv[0]
              << " <ip> open|close|width <w>|grasp <w>|homing|read [--speed v] [--force f]\n";
    return 1;
  }
  std::string ip = argv[1];
  std::string cmd = argv[2];
  double speed = 0.1;
  double force = 60.0;   // 30 → 60 N continuous force (Franka Hand max 70 N continuous)
                          // UMI 3D compliant fingers need sustained force to keep the deformed fit
  for (int i = 3; i < argc; ++i) {
    std::string a = argv[i];
    if (a == "--speed" && i + 1 < argc) speed = std::stod(argv[++i]);
    else if (a == "--force" && i + 1 < argc) force = std::stod(argv[++i]);
  }
  try {
    franka::Gripper g(ip);
    auto state = g.readOnce();
    if (cmd == "read") {
      std::cout << "{\"width\":" << state.width << ",\"max_width\":" << state.max_width
                << ",\"is_grasped\":" << (state.is_grasped ? "true" : "false")
                << ",\"temperature\":" << state.temperature << "}\n";
      return 0;
    }
    if (cmd == "open") {
      g.move(state.max_width, speed);
    } else if (cmd == "close") {
      // Use grasp() for sustained force, not move(0)
      // move() is position control: it stops on contact with the object but does not keep pressing, so the UMI soft fingers spring back
      // grasp(target=0, speed, force, eps_inner, eps_outer) keeps applying force to maintain the grip
      // eps_inner/outer set large (max_width) so any final width counts as success → sustained force
      g.grasp(0.0, speed, force, state.max_width, state.max_width);
    } else if (cmd == "homing") {
      g.homing();
    } else if (cmd == "width" || cmd == "grasp") {
      if (argc < 4) { std::cerr << "missing width\n"; return 2; }
      double w = std::stod(argv[3]);
      if (cmd == "width") {
        g.move(w, speed);
      } else {
        g.grasp(w, speed, force);
      }
    } else {
      std::cerr << "unknown cmd: " << cmd << "\n";
      return 3;
    }
    return 0;
  } catch (const franka::Exception& e) {
    std::cerr << "franka::Exception: " << e.what() << "\n";
    return 4;
  }
}

File: rt-host/franky_tools/gripper_cmd.py
#!/usr/bin/env python
"""gripper_cmd.py — franky 版一次性夹爪命令。对标 teleop/gripper_cmd.cpp(1:1)。

⚠️ 需 Franka Hand 在位(PushT 推杆构型无夹爪时连接会报 FCI refused,与 C++ 原版一致)。
夹爪连接独立于机械臂 FCI motion control,可与 servo 共存。

用法:
  python gripper_cmd.py <ip> read
  python gripper_cmd.py <ip> open
  python gripper_cmd.py <ip> close [--force 60]
  python gripper_cmd.py <ip> width <w_m> [--speed 0.1]
  python gripper_cmd.py <ip> grasp <w_m> [--speed 0.1] [--force 60]
  python gripper_cmd.py <ip> homing
"""
import sys
import franky


def main():
    if len(sys.argv) < 3:
        print("Usage: gripper_cmd.py <ip> open|close|width <w>|grasp <w>|homing|read "
              "[--speed v] [--force f]", file=sys.stderr)
        return 1
    ip, cmd = sys.argv[1], sys.argv[2]
    speed, force = 0.1, 60.0   # 同 C++:60N 持续力(Franka Hand max 70N,UMI 软指需持续施力)
    rest = sys.argv[3:]
    pos = [a for a in rest if not a.startswith("--")]
    i = 0
    while i < len(rest):
        if rest[i] == "--speed": speed = float(rest[i+1]); i += 2
        elif rest[i] == "--force": force = float(rest[i+1]); i += 2
        else: i += 1
    try:
        g = franky.Gripper(ip)
        mw = g.max_width
        if cmd == "read":
            print(f'{{"width":{g.width},"max_width":{mw},"is_grasped":{str(g.is_grasped).lower()}}}')
        elif cmd == "open":
            g.open(speed)                                  # 张到 max
        elif cmd == "close":
            # 同 C++:grasp(0, ..., eps_inner=eps_outer=max_width) → 任意终宽视为成功,持续施力
            g.grasp(0.0, speed, force, epsilon_inner=mw, epsilon_outer=mw)
        elif cmd == "homing":
            g.homing()
        elif cmd in ("width", "grasp"):
            if not pos:
                print("missing width", file=sys.stderr); return 2
            w = float(pos[0])
            g.move(w, speed) if cmd == "width" else g.grasp(w, speed, force)
        else:
            print(f"unknown cmd: {cmd}", file=sys.stderr); return 3
        return 0
    except franky.GripperException as e:
        print(f"GripperException: {str(e)[:120]}", file=sys.stderr); return 4
    except Exception as e:
        print(f"{type(e).__name__}: {str(e)[:120]}", file=sys.stderr); return 4


if __name__ == "__main__":
    sys.exit(main())
